dependency-specification.rst

```diff
@@ -0,0 +1,324 @@
+
+And a rule for direct reference specifications::
+
+    url_req       = name [*WSP extras] urlspec [*WSP url-marker]
+
+Leading to the unified rule that can specify a dependency::
+
+    specification = name_req / location_req
+
+Whitespace
+----------
+
+Non line-breaking whitespace is optional and has no semantic meaning.
+
+Names
+-----
+
+Python distribution names are currently defined in PEP-345 [#pep345]_. Names
+act as the primary identifier for distributions. They are present in all
+dependency specifications, and are sufficient to be a specification on their
+own.
+
+Extras
+------
+
+An extra is an optional part of a distribution. Distributions can specify as
+many extras as they wish, and each extra results in the declaration of
+additional dependencies of the distribution **when** the extra is used in a
+dependency specification. For instance::
+
+    requests[security]
+
+Extras union in the dependencies they define with the dependencies of the
+distribution they are attached to. The example above would result in requests
+being installed, and requests own dependencies, and also any dependencies that
+are listed in the "security" extra of requests.
+
+If multiple extras are listed, all the dependencies are unioned together.
+
+Versions
+--------
+
+See PEP-440 [#pep440]_ for more detail on both version numbers and version
+comparisons. Version specifications limit the versions of a distribution that
+can be used. They only apply to distributions looked up by name, rather than
+via a URL. Version comparison are also used in the markers feature. The
+optional brackets around a version are present for compatibility with PEP-345
+[#pep345]_ but should not be generated, only accepted.
+
+Environment Markers
+-------------------
+
+Environment markers allow a dependency specification to provide a rule that
+describes when the dependency should be used. For instance, consider a package
+that needs argparse. In Python 2.7 argparse is always present. On older Python
+versions it has to be installed as a dependency. This can be expressed as so::
+
+    argparse;python_version<"2.7"
+
+A marker expression evalutes to either True or False. When it evaluates to
+False, the dependency specification should be ignored.
+
+The marker language is a subset of Python itself, chosen for the ability to
+safely evaluate it without running arbitrary code that could become a security
+vulnerability. Markers were first standardised in PEP-345 [#pep345]_. This PEP
+fixes some issues that were observed in the described in PEP-426 [#pep426]_.
+
+Comparisons in marker expressions are typed by the comparison operator.  The
+<marker-op> operators that are not in <version-cmp> perform the same as they
+do for strings in Python. The <version-cmp> operators use the PEP-440
+[#pep440]_ version comparison rules if both sides are valid versions. If
+either side is not a valid version, then the comparsion falls back to the same
```

Comment:
But what started this whole subthread is that you (quite rightly) want to add version comparison semantics on top of what we have now, and so the question is how exactly to bolt this on :-). And in the process we're certainly going to break backcompat, in that e.g. `python_version >= "2.7.10"` might return something different before and after.

I can see an argument for defining `in` on version-typed objects if needed for backcompat, but I think the overall idea of being clear about types is better than ad-hoc fallback on error.

It would definitely help if @dstufft is able to generate a list of all the real-world-usages-seen-on-PyPI though :-).